### PR TITLE
WIP Auth Context and Master Password

### PR DESCRIPTION
**Pull request by Kodylow**
## Release Notes

- **New Features**
	- Enhanced configuration structure for guardians and gateways, introducing unique identifiers.
	- Improved payload structure for adding guardians and gateways, now including an `id` field.
	- Introduced a new authentication context for managing passwords, improving state management.
	- Added new localization entries for "Setup State" to enhance user experience.
	- Implemented a custom hook for managing authentication passwords across the application.
	- Introduced a new component for entering a master password, enhancing security.
	- Added a modal for connecting services with improved user interface and functionality.
	- Enhanced the `EditServiceModal` with password management features, including visibility toggling.
	- Introduced a new `SetMasterPassword` component to facilitate master password setup.
	- Added new components: `NoConnectedServices`, `ServiceTable`, and `ViewServiceButton` for better service management.
	- Updated the `ServiceCheckApi` to refine response structure and improve password handling.
	- Improved password management and configuration handling in the `GatewayApi` and `GuardianApi` classes.
	- Introduced conditional rendering for master password entry in the home page component.
	- New entries added to localization for improved user guidance and error handling.
	- Added a new `fetchWithFibonacciBackoff` method for robust API call management.
	- Enhanced management of setup states based on unique identifiers derived from the URL.
	- Introduced `ServiceInfoDisplay` and `ServiceTable` components for better service management.

- **Bug Fixes**
	- Adjusted local storage management to ensure state is correctly tied to specific instances.

- **Refactor**
	- Streamlined data handling in the `GuardianApi` and `GatewayApi` classes for better clarity and performance.
	- Enhanced the `Gateway` and `Guardian` context providers to incorporate authentication details.
	- Updated the structure of various hooks to return more comprehensive data for guardians and gateways.
	- Improved password management in API calls, enhancing security and encapsulation.
	- Renamed and restructured state management functions for better clarity and usability.
	- Enhanced the structure of the `ServiceCheckResponse` interface for improved functionality.
	- Updated the `BasicSettingsForm` component to enforce stricter type requirements for the password prop.

### CODE REVIEW

File: apps/router/src/context/guardian/SetupContext.tsx [hunk 1]
```diff
@@ -123,7 +132,9 @@
   initServerStatus,
   children,
 }: SetupContextProviderProps) => {
+  const guardianId = useLocation().pathname.split('/')[2];
   const api = useGuardianApi();
+  const initialState = useInitialState(guardianId);
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/context/guardian/SetupContext.tsx:137]: context only every loaded while in guardian/id, all good

File: apps/router/src/api/GatewayApi.ts [hunk 2]
```diff
@@ -20,85 +20,72 @@
...
 
-    return fetch(`${this.baseUrl}/${api}`, {
+    const tempPassword = password || this.password;
+
+    return fetch(`${this.config.config.baseUrl}/${api}`, {
       method: 'POST',
       headers: {
         'Content-Type': 'application/json',
-        Authorization: `Bearer ${password}`,
+        Authorization: `Bearer ${tempPassword}`,
```
[the remark this answers is not in the document]
Comment (otech47) [reply apps/router/src/api/GatewayApi.ts:69]: fair, worth at least wrapping in a trycatch?

File: apps/router/src/api/GatewayApi.ts [hunk 3]
```diff
@@ -20,85 +20,72 @@
...
+    const tempPassword = password || this.password;
 
-    return fetch(`${this.baseUrl}/${api}`, {
+    return fetch(`${this.config.config.baseUrl}/${api}`, {
       method: 'GET',
       headers: {
         'Content-Type': 'application/json',
-        Authorization: `Bearer ${password}`,
+        Authorization: `Bearer ${tempPassword}`,
       },
```
[the remark this answers is not in the document]
Comment (otech47) [reply apps/router/src/api/GatewayApi.ts:89]: fair, worth at least wrapping in a trycatch?

**Review by otech47 — COMMENTED** [review]
spent 10-15 mins of quick review, huge PR but maybe with some testing instructions i'll have time this weekend to see

code only review looks okay, left some emoji acks/nacks on the rabbit's suggestions... probably worth tightening up some of the issues but it seems like largely the security premise of storing passwords in the browser is a necessary part of this feature... but would be worth considering long-term plans to audit this code for maybe some basic cors hijacking situations if this code is loaded into insecure browsers

File: apps/router/src/api/GatewayApi.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (otech47) [reply apps/router/src/api/GatewayApi.ts:69]: Quiet rabbit no mansplaining it is 2024

**Comment by Kodylow** [comment]
@otech47 @alexlwn123 This is ready for review now, there's exactly 1 bug I'm still trying to crush: When you're doing a setup with multiple guardians in the same UI, the websocket closing/reopening when you flip between them sometimes throws an auth error and during setup that'll fail. Been working it all day

File: apps/router/src/context/guardian/SetupContext.tsx [hunk 4]
```diff
@@ -63,12 +64,10 @@ function makeInitialState(loadFromStorage = true): SetupState {
   return state;
 }
 
-const initialState = makeInitialState();
-
 const reducer = (state: SetupState, action: SetupAction): SetupState => {
   switch (action.type) {
     case SETUP_ACTION_TYPE.SET_INITIAL_STATE:
-      return makeInitialState(false);
+      return { ...state };
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/context/guardian/SetupContext.tsx:71]: handled in provider

File: apps/router/src/context/gateway/GatewayContext.tsx [hunk 5]
```diff
@@ -65,8 +66,14 @@ export const GatewayContextProvider: React.FC<GatewayContextProviderProps> = ({
   const [state, dispatch] = useReducer(reducer, initialState);
   const location = useLocation();
   const gatewayId = location.pathname.split('/')[2];
+  const { getGatewayPassword } = useAuth();
   const config = useGatewayConfig(gatewayId);
-  const gatewayApi = useMemo(() => new GatewayApi(config), [config]);
+  const password = getGatewayPassword(gatewayId);
+  console.log('password', password);
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/context/gateway/GatewayContext.tsx]: outdated

File: apps/router/src/context/AppContext.tsx [hunk 6]
```diff
@@ -179,29 +173,30 @@ export const AppContextProvider: React.FC<AppContextProviderProps> = ({
...
   useEffect(() => {
-    const isGuardian = (service: Service): service is GuardianConfig =>
+    const isGuardian = (service: ServiceConfig) =>
       service.baseUrl.startsWith('ws');
-    const isGateway = (service: Service): service is GatewayConfig =>
+    const isGateway = (service: ServiceConfig) =>
       service.baseUrl.startsWith('http');
-    const addService = async (service: Service) => {
+    const addService = async (service: ServiceConfig) => {
+      console.log('service', service);
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/context/AppContext.tsx:181]: fixed

File: apps/router/src/context/gateway/GatewayContext.tsx [hunk 7]
```diff
@@ -65,8 +66,16 @@ export const GatewayContextProvider: React.FC<GatewayContextProviderProps> = ({
   const [state, dispatch] = useReducer(reducer, initialState);
   const location = useLocation();
   const gatewayId = location.pathname.split('/')[2];
+  const { getGatewayPassword } = useAuth();
   const config = useGatewayConfig(gatewayId);
-  const gatewayApi = useMemo(() => new GatewayApi(config), [config]);
+  const password = getGatewayPassword(gatewayId);
+  if (password !== null) {
+    state.needsAuth = false;
+  }
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/context/gateway/GatewayContext.tsx]: fixed

File: apps/router/src/context/AuthContext.tsx [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+      payload: string;
+    };
+
+const saveToLocalStorage = (state: AuthContextValue) => {
+  const { guardianPasswords, gatewayPasswords } = state;
+  localStorage.setItem(
+    'fedimint_ui_auth',
+    JSON.stringify({ guardianPasswords, gatewayPasswords })
+  );
+};
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/context/AuthContext.tsx]: uses encrypyted passwords now, was just for dev, resolved

File: apps/router/src/home/ServiceTable.tsx [hunk 9]
```diff
@@ -57,48 +88,64 @@
                   `${type.charAt(0).toUpperCase() + type.slice(1)} URL`
                 )}
               </Th>
-              <Th width='200px'>{t('home.actions', 'Actions')}</Th>
+              {/* <Th>{t('common.status')}</Th> */}
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/home/ServiceTable.tsx]: removed comments, resolved

File: apps/router/src/hooks/useMasterPassword.tsx [hunk 10]
```diff
@@ -0,0 +1,8 @@
+import { useContext } from 'react';
+import { AuthContext } from '../context/AuthContext';
+
+export const useMasterPassword = () => {
+  const { masterPassword, setMasterPassword, isMasterPasswordSet } =
+    useContext(AuthContext);
+  return { masterPassword, setMasterPassword, isMasterPasswordSet };
+};
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/hooks/useMasterPassword.tsx]: resolved

File: apps/router/src/home/connectServiceModal/ServiceInfoDisplay.tsx [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+      {serviceInfo.serviceType === 'gateway' && (
+        <Box>
+          <Text fontSize='lg' fontWeight='bold' display='inline'>
+            {t('home.connect-service-modal.service-name-label')}:{' '}
+          </Text>
+          <Text fontSize='lg' display='inline'>
+            {serviceInfo.serviceName}
+          </Text>
+        </Box>
+      )}
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/home/connectServiceModal/ServiceInfoDisplay.tsx]: nope

File: apps/router/src/home/connectServiceModal/ServiceInfoDisplay.tsx [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+        </Text>
+        <Badge colorScheme='blue' fontSize='md' px={2} py={1}>
+          {serviceInfo.status === 'AWAITING_PASSWORD'
+            ? 'SETUP'
+            : serviceInfo.status}
+        </Badge>
+      </Box>
+    </Flex>
+  );
+};
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/home/connectServiceModal/ServiceInfoDisplay.tsx]: done

File: apps/router/src/home/EnterMasterPassword.tsx [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+  const toast = useToast();
+
+  const handleUnlock = async () => {
+    if (test) {
+      try {
+        const decryptedTest = await decrypt(test, password);
+        if (decryptedTest !== 'test') {
+          throw new Error('Invalid password');
+        }
+        setMasterPassword(decryptedTest);
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/home/EnterMasterPassword.tsx]: removing all crypto this side, moving to using browser password storage only

File: apps/router/src/context/guardian/GuardianContext.tsx [hunk 14]
```diff
@@ -59,8 +60,15 @@ export const GuardianContextProvider: React.FC<
...
+  const { decryptedServicePassword } = useUnlockedService(
+    guardianId,
+    'guardian'
+  );
   const config = useGuardianConfig(guardianId);
-  const guardianApi = useMemo(() => new GuardianApi(config), [config]);
+  const guardianApi = useMemo(
+    () => new GuardianApi(config, decryptedServicePassword),
+    [config, decryptedServicePassword]
+  );
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/context/guardian/GuardianContext.tsx]: unneeded

File: apps/router/src/home/connectServiceModal/useConnectServiceForm.tsx [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+    try {
+      const id = await sha256Hash(configUrl);
+      const serviceType = getServiceType(configUrl);
+
+      if (serviceType === 'guardian') {
+        dispatch({
+          type: APP_ACTION_TYPE.ADD_GUARDIAN,
+          payload: { id, guardian: { id, config: { baseUrl: configUrl } } },
+        });
+        storeGuardianPassword(id, password);
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/home/connectServiceModal/useConnectServiceForm.tsx]: this encrypts it, removing anyway, resolve

File: apps/router/src/hooks/useAuthContext.tsx [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+import { useMasterPassword } from './useMasterPassword';
+import { encrypt } from '../utils/crypto';
+
+export const useAuthContext = () => {
+  const { guardianPasswords, gatewayPasswords, dispatch } =
+    useContext(AuthContext);
+  const { masterPassword } = useMasterPassword();
+
+  const storeGuardianPassword = async (id: string, password: string) => {
+    if (!masterPassword) return;
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/hooks/useAuthContext.tsx:12]: done

File: apps/router/src/context/AuthContext.tsx [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+    }
+  }
+  return {
+    test: null,
+    guardianPasswords: {},
+    gatewayPasswords: {},
+    dispatch: () => null,
+    masterPassword: null,
+    setMasterPassword: () => null,
+    isMasterPasswordSet: false,
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/context/AuthContext.tsx]: needed, resolved

File: apps/router/src/context/AuthContext.tsx [hunk 18]
```diff
@@ -0,0 +1,166 @@
+import React, {
+  createContext,
+  Dispatch,
+  ReactNode,
+  useReducer,
+  useState,
+} from 'react';
+
+export interface AuthContextValue {
+  test: string | null;
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/context/AuthContext.tsx:12]: needed, resolved.

File: apps/router/src/home/services/ServiceTable.tsx [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+            requiresPassword: false,
+          };
+        }
+      }
+
+      // setServiceStatuses(statuses);
+    };
+
+    checkServices();
+  }, [services, type]);
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/home/services/ServiceTable.tsx]: fixed

File: apps/router/src/context/AuthContext.tsx [hunk 20]
```diff
@@ -0,0 +1,214 @@
...
+  guardianEncryptedPasswords: Record<string, string>;
+  gatewayEncryptedPasswords: Record<string, string>;
+  dispatch: Dispatch<AuthAction>;
+  masterPassword: string | null;
+  setMasterPassword: (password: string) => void;
+  isMasterPasswordSet: boolean;
+}
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/context/AuthContext.tsx:19]: fixed

File: apps/router/src/languages/ru.json [hunk 21]
```diff
@@ -510,15 +511,30 @@
...
+    "setup-state": "Состояние настройки",
+    "master-password": {
+      "master-password-explainer": "Установите мастер-пароль для пользовательского интерфейса, чтобы хранить зашифрованные учетные данные для каждого из ваших сервисов. Каждый сервис будет иметь свой собственный пароль при подключении, мастер-пароль используется только для этого интерфейса.",
+      "master-password": "Мастер-пароль",
+      "enter-master-password": "Введите мастер-пароль",
+      "create-master-password": "Создать мастер-пароль",
+      "confirm-master-password": "Подтвердите мастер-пароль",
+      "set-master-password": "Установить мастер-пароль",
+      "password-mismatch": "Пароли не совпадают",
+      "password-too-short": "Пароль должен быть не менее 8 символов в длину.",
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/languages/ru.json:535]: going to remove

File: apps/router/src/utils/crypto.ts [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+    },
+    key,
+    encoder.encode(plaintext)
+  );
+
+  const ciphertextArray = new Uint8Array(ciphertextBuffer);
+  const ciphertextBase64 = btoa(String.fromCharCode(...ciphertextArray));
+
+  return ciphertextBase64;
+}
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/utils/crypto.ts:44]: removing, using browser storage only now

File: apps/router/src/utils/crypto.ts [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+    {
+      name: 'AES-GCM',
+      iv: iv,
+    },
+    key,
+    ciphertextBytes
+  );
+
+  return decoder.decode(plaintextBuffer);
+}
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/utils/crypto.ts:91]: removing, using browser storage only now

File: apps/router/src/api/GuardianApi.ts [hunk 24]
```diff
@@ -26,23 +26,51 @@ export const SESSION_STORAGE_KEY = 'guardian-ui-key';
...
   private websocket: JsonRpcWebsocket | null = null;
   private connectPromise: Promise<JsonRpcWebsocket> | null = null;
+  private guardian: GuardianConfig;
+  private password: string | null;
 
-  constructor(private guardianConfig: GuardianConfig) {}
+  constructor(config: GuardianConfig, password: string | null) {
+    this.guardian = config;
+    this.password = password;
+  }
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/api/GuardianApi.ts:35]: separating out for auth manager refactor back to browser native password management

File: apps/router/src/home/HomePage.tsx [hunk 25]
```diff
@@ -19,40 +24,77 @@ export const HomePage: React.FC = () => {
...
+          </Heading>
+          {masterPassword ? (
+            <Button onClick={onConnectServiceOpen} colorScheme='blue'>
+              {t('home.connect-service-modal.label')}
+            </Button>
+          ) : (
+            <Button onClick={onSetMasterPasswordOpen} colorScheme='blue'>
+              {t('home.master-password.create-master-password')}
+            </Button>
+          )}
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/home/HomePage.tsx:60]: removing masterPassword scheme

File: apps/router/src/context/hooks.tsx [hunk 26]
```diff
@@ -39,14 +39,19 @@ import {
...
   return useContext(AppContext);
 }
 
 export function useAppGuardianConfigs(): GuardianConfig[] {
   const { guardians } = useAppContext();
-  return Object.values(guardians).map((guardian) => guardian.config);
+  return Object.values(guardians).map((guardian) => ({
+    id: guardian.id,
+    config: guardian.config,
+  }));
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/context/hooks.tsx:54]: unused, removed

File: apps/router/src/context/hooks.tsx [hunk 27]
```diff
@@ -61,14 +66,14 @@
   const { guardians } = useAppContext();
   if (!guardians[id])
     throw new Error('useGuardianConfig must be used with a selected guardian');
-  return guardians[id].config;
+  return { id, config: guardians[id].config };
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/context/hooks.tsx:69]: matches interface, resolved

File: apps/router/src/context/hooks.tsx [hunk 28]
```diff
@@ -61,14 +66,14 @@
...
 };
 
 export const useGatewayConfig = (id: string): GatewayConfig => {
   const { gateways } = useAppContext();
   if (!gateways[id])
     throw new Error('useGatewayConfig must be used with a selected gateway');
-  return gateways[id].config;
+  return { id, config: gateways[id].config };
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/context/hooks.tsx:76]: matches interface, resolved

File: apps/router/src/guardian-ui/components/setup/screens/setConfiguration/SetConfiguration.tsx [hunk 29]
```diff
@@ -191,6 +190,7 @@ export const SetConfiguration: React.FC<Props> = ({ next }: Props) => {
       if (isHost || isSolo) {
         // Hosts set their own connection name
         // - They should submit both their local and the consensus config gen params.
+        console.log('submitConfig', password);
```
[the remark this answers is not in the document]
Comment (Kodylow) [reply apps/router/src/guardian-ui/components/setup/screens/setConfiguration/SetConfiguration.tsx:193]: removed

**Comment by Kodylow** [comment]
Closing in favor of infinitely simpler version:

https://github.com/fedimint/ui/pull/558

**Pull request closed without merging**